File: codes/java/chap10/answer_A77.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 입력
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int L = sc.nextInt();
		int K = sc.nextInt();
		int[] A = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			A[i] = sc.nextInt();
		}

		// 바이너리 서치(left: 현재 하한선 / right: 현재 상한선)
		int left = 1, right = 1000000000;
		while (left < right) {
			int mid = (left + right + 1) / 2;
			boolean answer = check(N, L, K, A, mid);
			if (answer == false) {
				right = mid - 1; // 답이 전반 부분으로 좁혀진다
			}
			else {
				left = mid;      // 답이 후반 부분으로 좁혀진다
			}
		}

		// 출력
		System.out.println(left);
	}

	// 점수의 최대값이 x 이상인지 여부를 판단하는 함수
	static boolean check(int N, int L, int K, int[] A, int x) {
		int count = 0;      // 현재 몇 번을 끊었는지를 나타낸다
		int lastKireme = 0; // 마지막에 자른 곳을 표시한다
		for (int i = 1; i <= N; i++) {
			if (A[i] - lastKireme >= x && L - A[i] >= x) {
				count += 1;
				lastKireme = A[i];
			}
		}
		return (count >= K);
	}
}

File: codes/cpp/chap10/answer_A77.cpp
#include <iostream>
using namespace std;

int N, L, K, A[100009];

bool check(int x) {
	int Count = 0, Last_Kireme = 0;// Count는 현 시점에서 몇 번 잘랐는가를 나타낸다
	for (int i = 1; i <= N; i++) {
		if (A[i] - Last_Kireme >= x && L - A[i] >= x) {
			Count += 1;
			Last_Kireme = A[i];
		}
	}
	if (Count >= K) return true;
	return false;
}

int main() {
	// 입력
	cin >> N >> L >> K;
	for (int i = 1; i <= N; i++) cin >> A[i];

	// 바이너리 서치(left: 현재의 하한/ right: 현재의 상한)
	long long left = 1, right = 1'000'000'000;
	while (left < right) {
		long long mid = (left + right + 1) / 2;
		bool Answer = check(mid);
		if (Answer == false) right = mid - 1; // 답이 전반 부분으로 좁혀진다
		if (Answer == true) left = mid; // 답이 후반 부분으로 좁혀진다
	}

	// 출력
	cout << left << endl;
	return 0;
}
